src/managed_process.rs

```diff
@@ -26,7 +26,53 @@ fn seconds_since_epoch() -> f64 {
 pub struct ManagedProcess {
     kill_signal: Arc<Mutex<u32>>,
     pid:         Arc<Mutex<Option<u32>>>,
-    thread:      JoinHandle<()>
+    thread:      JoinHandle<()>,
+    backoff:     Arc<RwLock<Backoff>>,
+}
+
+struct Backoff {
+    restart_count: u64,
+    restart_threshold: f64,
+    start_time: f64,
+    end_time: f64,
+    backoff: u64,
+}
+
+impl Backoff {
+
+    fn new(restart_threshold: f64) -> Self {
+        Backoff {
+            start_time: restart_threshold,
+            restart_threshold: restart_threshold,
+            end_time: 0.0_f64,
+            backoff: 0,
+            restart_count: 0,
+        }
+    }
+
+    fn next_backoff(&mut self) -> Duration {
+        self.end_time = seconds_since_epoch();
+
+        let seconds_to_backoff =
+            if (self.end_time - self.start_time) < self.restart_threshold {
+                self.backoff += 1;
+
+                // non-linear back off
+                (self.backoff * self.backoff) >> 1
```

Comment:
That will be 0 at the first call of `next_backoff()`, when `self.backoff == 1` and thus we restart the process immediately. I feel we may want to initialize `self.backoff` to 1 instead.